File: miniprogram/src/services/manual.ts
import { apiRequest } from './api'
import { agentAbilityMap, agentOptions, mapOptions, throwOptions } from './labels'
import type { Lineup } from './types'

export type LineupBaseValue = {
  map: string
  site: string
  side: string
  agent: string
  ability: string
  throw_type: string
}

export type ManualLineupFormPayload = LineupBaseValue & {
  standing_description: string
  aim_description: string
  landing_description: string
  minimap_x?: number | null
  minimap_y?: number | null
  landing_x?: number | null
  landing_y?: number | null
}

export type ManualImagePayload = {
  image_base64: string
  filename?: string
  note: string
}

export type ManualUploadPayload = {
  form: ManualLineupFormPayload
  images: ManualImagePayload[]
}

export type ExternalVideoResolveResponse = {
  playable_url: string
  title?: string | null
  duration_seconds?: number | null
}

export type FrameNodePayload = {
  timestamp_ms: number
  note: string
  order_index: number
}

export type VideoFrameSubmitPayload = {
  source_url?: string | null
  timestamps: {
    standing: number
    aim: number
    landing: number
  }
  form: ManualLineupFormPayload
  frame_nodes: FrameNodePayload[]
}

export type CorrectionSession = {
  video_url?: string | null
  original_timestamps: {
    standing: number
    aim: number
    landing: number
  }
  original_form: ManualLineupFormPayload
}

export function defaultLineupBaseValue(): LineupBaseValue {
  const agent = agentOptions[0]?.value || 'sova'
  return {
    map: mapOptions[0]?.value || 'ascent',
    site: 'a',
    side: 'attack',
    agent,
    ability: agentAbilityMap[agent]?.[0]?.value || '',
    throw_type: throwOptions[0]?.value || 'direct'
  }
}

export function withDefaultDescriptions(
  value: LineupBaseValue,
  descriptions?: Partial<Pick<ManualLineupFormPayload, 'standing_description' | 'aim_description' | 'landing_description'>>
): ManualLineupFormPayload {
  return {
    ...value,
    standing_description: descriptions?.standing_description || '',
    aim_description: descriptions?.aim_description || '',
    landing_description: descriptions?.landing_description || ''
  }
}

export async function uploadManualLineup(payload: ManualUploadPayload): Promise<Lineup> {
  return apiRequest<Lineup>('/lineups/manual-upload-json', {
    method: 'POST',
    data: payload
  })
}

export async function resolveExternalVideo(sourceUrl: string): Promise<ExternalVideoResolveResponse> {
  return apiRequest<ExternalVideoResolveResponse>('/manual/external-video/resolve', {
    method: 'POST',
    data: { source_url: sourceUrl }
  })
}

export async function submitManualVideo(payload: VideoFrameSubmitPayload): Promise<Lineup> {
  return apiRequest<Lineup>('/manual/video/submit', {
    method: 'POST',
    data: payload
  })
}

export async function getCorrectionSession(lineupId: number): Promise<CorrectionSession> {
  return apiRequest<CorrectionSession>(`/lineups/${lineupId}/correction-session`, { method: 'POST' })
}

export async function submitCorrection(lineupId: number, payload: VideoFrameSubmitPayload): Promise<Lineup> {
  return apiRequest<Lineup>(`/lineups/${lineupId}/corrections`, {
    method: 'POST',
    data: payload
  })
}

File: backend/app/schemas/manual.py
from pydantic import BaseModel, Field

from app.enums import MapName, Side, ThrowType


class VideoSessionCreate(BaseModel):
    source_url: str | None = Field(default=None, max_length=2000)


class ExternalVideoResolveRequest(BaseModel):
    source_url: str = Field(min_length=1, max_length=2000)


class ExternalVideoResolveResponse(BaseModel):
    playable_url: str
    title: str | None = None
    duration_seconds: float | None = None


class FrameTimestamps(BaseModel):
    standing: int = Field(ge=0)
    aim: int = Field(ge=0)
    landing: int = Field(ge=0)


class FrameNodeSubmit(BaseModel):
    timestamp_ms: int = Field(ge=0)
    note: str = ""
    order_index: int = Field(ge=0)


class ManualLineupForm(BaseModel):
    map: MapName
    agent: str
    side: Side
    ability: str
    throw_type: ThrowType
    site: str = "a"
    standing_description: str = ""
    aim_description: str = ""
    landing_description: str = ""
    minimap_x: float | None = Field(default=None, ge=0.0, le=1.0)
    minimap_y: float | None = Field(default=None, ge=0.0, le=1.0)
    landing_x: float | None = Field(default=None, ge=0.0, le=1.0)
    landing_y: float | None = Field(default=None, ge=0.0, le=1.0)


class ManualImageSubmit(BaseModel):
    image_base64: str = Field(min_length=1)
    filename: str | None = Field(default=None, max_length=255)
    note: str = Field(default="", max_length=1200)


class ManualJsonUploadRequest(BaseModel):
    form: ManualLineupForm
    images: list[ManualImageSubmit] = Field(min_length=1, max_length=20)


class VideoFrameSubmit(BaseModel):
    source_url: str | None = Field(default=None, max_length=2000)
    timestamps: FrameTimestamps
    form: ManualLineupForm
    frame_nodes: list[FrameNodeSubmit] = Field(default_factory=list, max_length=20)


class CorrectionSessionResponse(BaseModel):
    video_url: str | None
    original_timestamps: FrameTimestamps
    original_form: ManualLineupForm
